src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.InFlightSolution.cs

```diff
@@ -80,18 +105,29 @@ public void TryKickOffPrimaryBranchWork_NoLock(Func<Solution, CancellationToken,
             {
                 Contract.ThrowIfFalse(_workspace._gate.CurrentCount == 0);
                 Contract.ThrowIfNull(updatePrimaryBranchAsync);
+                Contract.ThrowIfTrue(this._cancellationTokenSource_doNotAccessDirectly.IsCancellationRequested);
 
                 // Already set up the work to update the primary branch
                 if (_primaryBranchTask != null)
                     return;
 
-                _primaryBranchTask = ComputePrimaryBranchAsync();
+                // Grab the cancellation token up front while we know we are holding the lock and mutating state.  We
+                // don't want to potentially grab it at some undefined point at the future when this type has already
+                // had its in-flight-count reduced to 0 and thus had our CTS be disposed.
+                //
+                // Also kick this off in a dedicated task.  The state mutation updating of this InFlightSolution and the
+                // cache state in RemoteWorkspace must always run fully to completion without issue. We don't want
+                // anything we call in this method to possibly prevent the method from running to completion.
+                var cancellationToken = this.CancellationToken;
+                _primaryBranchTask = Task.Run(() => ComputePrimaryBranchAsync(cancellationToken), cancellationToken);
```

Comment:
same issue.  TryKickOffPrimaryBranchWork_NoLock must run to completion successfully.